pyuvsim/simsetup.py

```diff
@@ -285,29 +287,199 @@ def create_mock_catalog(time, arrangement='zenith', array_location=None, Nsrcs=N
     stokes = np.zeros((4, 1, Nsrcs))
     stokes[0, :] = fluxes
     catalog = pyradiosky.SkyModel(names, ra, dec, stokes, 'flat')
-    if return_table:
-        return pyradiosky.skymodel_to_array(catalog), mock_keywords
+    if return_data:
+        catalog = SkyModelData(catalog)
     if get_rank() == 0 and save:
         np.savez('mock_catalog_' + arrangement, ra=ra.rad, dec=dec.rad, alts=alts, azs=azs,
                  fluxes=fluxes)
 
     return catalog, mock_keywords
 
 
+class SkyModelData:
+    """
+    Carries immutable SkyModel data in simple ndarrays.
+
+    This is to facilitate sharing SkyModel objects in MPI, without
+    excessive copying and memory bloat.
+
+    When used with MPI, this can be initialized simultaneously on all
+    processes such that sky_in is provided only on the root process.
+    The data can then be shared across all processes by running the `share`
+    method.
+
+    Parameters
+    ----------
+    sky_in: :class:~`pyradiosky.SkyModel`
+        A valid SkyModel object.
+    """
+
+    spectral_type = None
+    Ncomponents = None
+    ra = None
+    dec = None
+    name = None
+    Nfreqs = None
+    stokes_I = None
+    stokes_Q = None
+    stokes_U = None
+    stokes_V = None
+    freq_array = None
+    reference_frequency = None
+    spectral_index = None
+    polarized = None
+
+    ncomp_attrs = ['stokes_I', 'stokes_Q', 'stokes_U', 'stokes_V',
+                   'ra', 'dec', 'reference_frequency', 'spectral_index', 'name']
+
+    def __init__(self, sky_in=None):
+        # Collect relevant attributes.
+        if sky_in is not None:
+            self.name = sky_in.name
+            self.spectral_type = sky_in.spectral_type
+            self.Ncomponents = sky_in.Ncomponents
+            self.ra = sky_in.ra.deg
+            self.dec = sky_in.dec.deg
+            self.Nfreqs = sky_in.Nfreqs
+            self.stokes_I = sky_in.stokes[0, ...]
+
+            self.polarized = sky_in._polarized
+            if sky_in._n_polarized > 0:
+                Q, U, V = sky_in.stokes[1:, :, sky_in._polarized]
+                self.stokes_Q = Q
+                self.stokes_U = U
+                self.stokes_V = V
+
+            if sky_in._freq_array.required:
+                self.freq_array = sky_in.freq_array.to("Hz").value
+            if sky_in._reference_frequency.required:
+                self.reference_frequency = sky_in.reference_frequency.to("Hz").value
+            if sky_in._spectral_index.required:
+                self.spectral_index = sky_in.spectral_index
+
+    def __getitem__(self, inds):
```

Comment:
After reading some about `__getitem__` this seems to be a non-standard implementation -- it's returning a downselected copy of the object which is not what `__getitem__` is generally used for.